File: content-solutions/bfs/safe-dial.py
from collections import deque


def solution(target, jammed):
    blocked = set(jammed)
    if "0000" in blocked:
        return -1
    dist = {"0000": 0}
    queue = deque(["0000"])
    while queue:
        s = queue.popleft()
        if s == target:
            return dist[s]
        for i in range(4):
            for d in (1, -1):
                digit = (int(s[i]) + d) % 10
                t = s[:i] + str(digit) + s[i + 1:]
                if t not in blocked and t not in dist:
                    dist[t] = dist[s] + 1
                    queue.append(t)
    return -1

File: content-solutions/bfs/safe-dial.java
import java.util.*;

class Solution {
    public int solution(String target, String[] jammed) {
        Set<String> blocked = new HashSet<>(Arrays.asList(jammed));
        if (blocked.contains("0000")) return -1;
        Map<String, Integer> dist = new HashMap<>();
        dist.put("0000", 0);
        Queue<String> queue = new ArrayDeque<>();
        queue.offer("0000");
        while (!queue.isEmpty()) {
            String s = queue.poll();
            if (s.equals(target)) return dist.get(s);
            for (int i = 0; i < 4; i++) {
                for (int d : new int[] {1, 9}) {
                    int digit = (s.charAt(i) - '0' + d) % 10;
                    String t = s.substring(0, i) + digit + s.substring(i + 1);
                    if (!blocked.contains(t) && !dist.containsKey(t)) {
                        dist.put(t, dist.get(s) + 1);
                        queue.offer(t);
                    }
                }
            }
        }
        return -1;
    }
}
